src/helpers/synthetic_peers.rs

```diff
@@ -0,0 +1,180 @@
+use crate::protocol::{
+    message::{constants::HEADER_LEN, filter::MessageFilter, Message},
+    payload::{codec::Codec, Version},
+};
+
+use assert_matches::assert_matches;
+use pea2pea::{
+    connections::ConnectionSide,
+    protocols::{Handshaking, Reading, Writing},
+    Connection, Node, Pea2Pea,
+};
+use tokio::sync::mpsc::{self, Receiver, Sender};
+
+use std::{
+    io::{Cursor, Result},
+    net::SocketAddr,
+};
+
+pub struct SyntheticNode {
+    inner_node: InnerNode,
+    inbound_rx: Receiver<Message>,
+}
+
+impl SyntheticNode {
+    pub fn new(node: Node, enable_handshaking: bool, message_filter: MessageFilter) -> Self {
+        // Inbound channel size of 100 messages.
+        let (tx, rx) = mpsc::channel(100);
+        let inner_node = InnerNode::new(node, tx, message_filter);
+
+        // Enable the read and write protocols, handshake is enabled on a per-case basis.
+        inner_node.enable_reading();
+        inner_node.enable_writing();
+
+        if enable_handshaking {
+            inner_node.enable_handshaking();
+        }
+
+        Self {
+            inner_node,
+            inbound_rx: rx,
+        }
+    }
+
+    pub async fn connect(&self, target: SocketAddr) -> Result<()> {
```

Comment:
An alternate API to consider:

1. Have a SynthNodeBuilder (which has no recv/send methods)
2. Has a variety of `build` type methods, which return a `SynthNode`
     1. `connect` (connect, without any handshaking)
     2. `accept` (listen and accept without handshaking)
     3. `initiate_handshake` (connect and handshake)
     4. `recieve_handshake` (listen, accept and handshake)
     5. ... 

This prevents forgetting to connect (since one cannot receive / send on the builder). But could be too much boiler plate for not much more upside (:

Reply:
Or did you mean as a replacement for `SyntheticNode::new`?

Reply:
Yes, as a replacement for `new`. All cases are currently covered, but one can misuse by doing

```
let peer = SyntheticNode ::new(...);
peer.recv_message(); // will panic!("all receivers dropped");
```

I'm suggesting something along these lines

```
let peer = SyntheticNode::builder(...)
                    .initiate_handshake(node_addr).await.unwrap();
peer.recv_message();
```

but its mostly a trivial thing, just means you can't forget to call `connect` etc.

Reply:
I'd actually expect that first example to simply hang, since it will just `await` on the channel `recv`.